<!DOCTYPE html>
<html>
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
        <title>Graceful exit with cluster and pm</title>
        <meta name="viewport" content="width=device-width">

        <!-- syntax highlighting CSS -->
        <link rel="stylesheet" href="/google_reader/css/syntax.css">

        <!-- Custom CSS -->
        <link rel="stylesheet" href="/google_reader/css/main.css">

    </head>
    <body>

        <div class="container">
          <div class="site">
            <div class="header">
              <h1 class="title"><a href="/google_reader/">Google Reader</a></h1>
              <a class="extra" href="/google_reader/">home</a>
            </div>

                <h2>Graceful exit with cluster and pm</h2>
<p class="meta">2013-06-07 19:20</p>

<div class="post">
<h2>Graceful exit with cluster and pm</h2>

<h3>by MK2</h3>

<h3>at 2013-06-07 11:20:00</h3>

<h3>original <a href="http://www.cnblogs.com/fengmk2/archive/2013/06/07/graceful-exit-pm-cluster.html">http://www.cnblogs.com/fengmk2/archive/2013/06/07/graceful-exit-pm-cluster.html</a></h3>

<p>相信大家看完 <a href="https://github.com/dead-horse">@不四</a> 的 <a href="http://deadhorse.me/nodejs/2013/04/13/exception_and_domain.html">Node.js 异步异常的处理与domain模块解析</a> 后，已经对<a href="http://nodejs.org/docs/latest/api/domain.html">domain</a> 模块有了更深一步了解了。</p>


<p><br><br><p>本次分享主要介绍一下使用 <a href="https://github.com/fengmk2/graceful">graceful</a> 配合 <a href="http://nodejs.org/docs/latest/api/cluster.html">cluster</a> 和 <a href="https://github.com/aleafs/pm">pm</a> 实现当发生 <code>uncaughtException</code> 的时候优雅退出(graceful exit)。</p><br><br><h2>Cluster</h2><br><br><p>官方<a href="http://nodejs.org/docs/latest/api/cluster.html">cluster</a>模块的例子原型代码来自: <a href="http://nodejs.org/docs/latest/api/domain.html#domain_warning_don_t_ignore_errors">Warning: Don’t Ignore Errors!</a></p><br><br><p><a href="https://github.com/fengmk2/graceful">graceful</a> 的默认示例是基于 <a href="http://nodejs.org/docs/latest/api/cluster.html">cluster</a> 写的: <a href="https://github.com/fengmk2/graceful/tree/master/example/connect_with_cluster">connect with cluster example</a></p><br><br><p>一般地，我会将代码结构分为3个文件: </p><br><br><ul><br><li><code>dispatch.js</code>: master, 管理worker进程，报警等</li><br><li><code>worker.js</code>: worker，实际工作进程的容器</li><br><li><code>app.js</code>: web server, hsf, other services and other logic.</li><br></ul><br><br><h3><a href="https://github.com/fengmk2/graceful/blob/master/example/connect_with_cluster/dispatch.js">dispatch.js</a></h3><br><br><pre><code>var cluster = require(&#39;cluster&#39;);<br>var path = require(&#39;path&#39;);<br><br>cluster.setupMaster({<br>  exec: path.join(<strong>dirname, &#39;worker.js&#39;)<br>});<br><br>cluster.fork();<br>cluster.fork();<br><br>cluster.on(&#39;disconnect&#39;, function (worker) {<br>  var w = cluster.fork();<br>  console.error(&#39;[%s] [master:%s] wroker:%s disconnect! new worker:%s fork&#39;, <br>    new Date(), process.pid, worker.process.pid, w.process.pid);<br>});<br><br>cluster.on(&#39;exit&#39;, function (worker) {<br>  console.error(&#39;[%s] [master:%s] wroker:%s exit!&#39;, <br>    new Date(), process.pid, worker.process.pid);<br>});</code></pre><br><br><h3><a href="https://github.com/fengmk2/graceful/blob/master/example/connect_with_cluster/app.js">app.js</a></h3><br><br><p>app.js 不会涉及 <a href="http://nodejs.org/docs/latest/api/cluster.html">cluster</a> 和 <a href="https://github.com/fengmk2/graceful">graceful</a> 的代码，让业务逻辑不需要关注系统的基本功能。</p><br><br><pre><code>var http = require(&#39;http&#39;);<br>var connect = require(&#39;connect&#39;);<br><br>var app = connect(<br>  function (req, res, next) {<br>    req.on(&#39;end&#39;, function () {<br>      if (req.url === &#39;/asycerror&#39;) {<br>        setTimeout(function () {<br>          foo.bar();<br>        }, 10);<br>        return;<br>      }<br>      process.nextTick(function () {<br>        res.setHeader(&#39;content-type&#39;, &#39;text/json&#39;);<br>        res.end(JSON.stringify({<br>          method: req.method,<br>          url: req.url,<br>          headers: req.headers,<br>          Connection: res.getHeader(&#39;connection&#39;) || &#39;keep-alive&#39;,<br>          pid: process.pid,<br>        }));<br>      });<br>    });<br>    req.resume();<br>  },<br>  function (err, req, res, next) {<br>    var domainThrown = err.domain_thrown || err.domainThrown;<br>    var msg = &#39;domainThrown: &#39; + domainThrown + &#39;\n&#39; + err.stack;<br>    console.error(&#39;%s %s\n%s&#39;, req.method, req.url, msg);<br>    res.statusCode = 500;<br>    res.setHeader(&#39;content-type&#39;, &#39;text/plain&#39;);<br>    res.end(msg + &#39;\n&#39;);<br>  }<br>);<br><br>var server = http.createServer(app);<br>module.exports = server;</code></pre><br><br><h3><a href="https://github.com/fengmk2/graceful/blob/master/example/connect_with_cluster/worker.js">worker.js</a></h3><br><br><pre><code>var PORT = +process.env.PORT || 1337;<br>var graceful = require(&#39;graceful&#39;);<br>var server = require(&#39;./app&#39;);<br>server.listen(PORT);<br>console.log(&#39;[%s] [worker:%s] web server start listen on %s&#39;, new Date(), process.pid, PORT);<br><br>var restapi = require(&#39;http&#39;).createServer().listen(1985);<br>console.log(&#39;[%s] [worker:%s] rest api start listen on %s&#39;, new Date(), process.pid, 1985);<br><br>graceful({<br>  server: [server, restapi],<br>  killTimeout: 10000,<br>  error: function (err, throwErrorCount) {<br>    // you can do custom log here, send email, call phone and so on...<br>    if (err.message) {<br>      err.message += &#39; (uncaughtException throw &#39; + throwErrorCount + &#39; times on pid:&#39; + process.pid + &#39;)&#39;;<br>    }<br>    // logger.error(err);<br>  }<br>});</code></pre><br><br><h3>stdout</h3><br><br><pre><code>$ node example/connect_with_cluster/dispatch.js <br>[Fri Apr 19 2013 09:32:29 GMT+0800 (CST)] [worker:10231] web server start listen on 1337<br>[Fri Apr 19 2013 09:32:29 GMT+0800 (CST)] [worker:10231] rest api start listen on 1985<br>[Fri Apr 19 2013 09:32:29 GMT+0800 (CST)] [worker:10232] web server start listen on 1337<br>[Fri Apr 19 2013 09:32:29 GMT+0800 (CST)] [worker:10232] rest api start listen on 1985<br><br>$ curl localhost:1337/<br>{&quot;method&quot;:&quot;GET&quot;,&quot;url&quot;:&quot;/&quot;,&quot;headers&quot;:{&quot;user-agent&quot;:&quot;curl/7.24.0 (x86_64-apple-darwin12.0) libcurl/7.24.0 OpenSSL/0.9.8r zlib/1.2.5&quot;,&quot;host&quot;:&quot;localhost:1337&quot;,&quot;accept&quot;:&quot;<em>/</em>&quot;},&quot;Connection&quot;:&quot;keep-alive&quot;,&quot;pid&quot;:10232}<br><br>$ curl localhost:1337/asycerror<br># hang for a while<br>curl: (52) Empty reply from server<br><br># Master<br>$ node example/connect_with_cluster/dispatch.js <br>[Fri Apr 19 2013 09:32:29 GMT+0800 (CST)] [worker:10231] web server start listen on 1337<br>[Fri Apr 19 2013 09:32:29 GMT+0800 (CST)] [worker:10231] rest api start listen on 1985<br>[Fri Apr 19 2013 09:32:29 GMT+0800 (CST)] [worker:10232] web server start listen on 1337<br>[Fri Apr 19 2013 09:32:29 GMT+0800 (CST)] [worker:10232] rest api start listen on 1985<br>[uncaughtException] throw error 1 times<br>[ReferenceError: foo is not defined (uncaughtException throw 1 times on pid:10232)]<br>[Fri Apr 19 2013 09:34:21 GMT+0800 (CST)] [worker:10232] close 2 servers!<br>[Fri Apr 19 2013 09:34:21 GMT+0800 (CST)] [worker:10232] worker disconnect!<br>[Fri Apr 19 2013 09:34:21 GMT+0800 (CST)] [master:10230] wroker:10232 disconnect! new worker:10288 fork<br>[Fri Apr 19 2013 09:34:21 GMT+0800 (CST)] [worker:10288] web server start listen on 1337<br>[Fri Apr 19 2013 09:34:21 GMT+0800 (CST)] [worker:10288] rest api start listen on 1985<br>[Fri Apr 19 2013 09:34:31 GMT+0800 (CST)] [worker:10232] kill timeout, exit now.<br>[Fri Apr 19 2013 09:34:31 GMT+0800 (CST)] [master:10230] wroker:10232 exit!<br><br># test again<br>$ curl localhost:1337/asycerror<br># hang for a while<br>curl: (52) Empty reply from server<br><br># Master<br>$ node example/connect_with_cluster/dispatch.js <br>[Fri Apr 19 2013 09:32:29 GMT+0800 (CST)] [worker:10231] web server start listen on 1337<br>[Fri Apr 19 2013 09:32:29 GMT+0800 (CST)] [worker:10231] rest api start listen on 1985<br>[Fri Apr 19 2013 09:32:29 GMT+0800 (CST)] [worker:10232] web server start listen on 1337<br>[Fri Apr 19 2013 09:32:29 GMT+0800 (CST)] [worker:10232] rest api start listen on 1985<br>[uncaughtException] throw error 1 times<br>[ReferenceError: foo is not defined (uncaughtException throw 1 times on pid:10232)]<br>[Fri Apr 19 2013 09:34:21 GMT+0800 (CST)] [worker:10232] close 2 servers!<br>[Fri Apr 19 2013 09:34:21 GMT+0800 (CST)] [worker:10232] worker disconnect!<br>[Fri Apr 19 2013 09:34:21 GMT+0800 (CST)] [master:10230] wroker:10232 disconnect! new worker:10288 fork<br>[Fri Apr 19 2013 09:34:21 GMT+0800 (CST)] [worker:10288] web server start listen on 1337<br>[Fri Apr 19 2013 09:34:21 GMT+0800 (CST)] [worker:10288] rest api start listen on 1985<br>[Fri Apr 19 2013 09:34:31 GMT+0800 (CST)] [worker:10232] kill timeout, exit now.<br>[Fri Apr 19 2013 09:34:31 GMT+0800 (CST)] [master:10230] wroker:10232 exit!<br><br>[uncaughtException] throw error 1 times<br>[ReferenceError: foo is not defined (uncaughtException throw 1 times on pid:10288)]<br>[Fri Apr 19 2013 09:36:17 GMT+0800 (CST)] [worker:10288] close 2 servers!<br>[Fri Apr 19 2013 09:36:17 GMT+0800 (CST)] [worker:10288] worker disconnect!<br>[Fri Apr 19 2013 09:36:17 GMT+0800 (CST)] [master:10230] wroker:10288 disconnect! new worker:10351 fork<br>[Fri Apr 19 2013 09:36:17 GMT+0800 (CST)] [worker:10351] web server start listen on 1337<br>[Fri Apr 19 2013 09:36:17 GMT+0800 (CST)] [worker:10351] rest api start listen on 1985<br>[Fri Apr 19 2013 09:36:27 GMT+0800 (CST)] [worker:10288] kill timeout, exit now.<br>[Fri Apr 19 2013 09:36:27 GMT+0800 (CST)] [master:10230] wroker:10288 exit!</code></pre><br><br><h2>Graceful exit with</h2><br><br><p><a href="https://github.com/aleafs/pm">pm</a> A graceful node library to contribute a permanent “master-worker” server.</p><br><br><p>同样，也是按照3文件划分职能。<a href="https://github.com/aleafs/pm/tree/master/demo/graceful_exit">graceful exit demo</a></p><br><br><h3><a href="https://github.com/aleafs/pm/blob/master/demo/graceful_exit/dispatch.js">dispatch.js</a></h3><br><br><pre><code>var master = require(&#39;pm&#39;).createMaster();<br><br>master.on(&#39;giveup&#39;, function (name, fatals, pause) {<br>  console.log(&#39;[%s] [master:%s] giveup to restart &quot;%s&quot; process after %d times. pm will try after %d ms.&#39;, <br>    new Date(), process.pid, name, fatals, pause);<br>});<br><br>master.on(&#39;disconnect&#39;, function (name, pid) {<br>  // console.log(&#39;%s %s disconnect&#39;, name, pid)<br>  var w = master.fork(name);<br>  console.error(&#39;[%s] [master:%s] worker:%s disconnect! new worker:%s fork&#39;, <br>    new Date(), process.pid, pid, w.process.pid);<br>});<br><br>master.on(&#39;fork&#39;, function (name, pid) {<br>  console.log(&#39;[%s] [master:%s] new %s:worker:%s fork&#39;,<br>    new Date(), process.pid, name, pid);<br>});<br><br>master.on(&#39;quit&#39;, function (name, pid, code, signal) {<br>  console.log(&#39;[%s] [master:%s] %s:worker:%s quit, code: %s, signal: %s&#39;,<br>    new Date(), process.pid, name, pid, code, signal);<br>});<br><br>master.register(&#39;web&#39;, </strong>dirname + &#39;/worker.js&#39;, {<br>  listen: 1984,<br>  children: 2<br>});<br><br>master.dispatch();</code></pre><br><br><h3><a href="http://www.cnblogs.com/fengmk2/">app.js</a></h3><br><br><pre><code>var http = require(&#39;http&#39;);<br><br>var server = http.createServer(function (req, res) {<br>  if (req.url === &#39;/asyncerror&#39;) {<br>    setTimeout(function () {<br>      asyncError();<br>    }, 10);<br>    return;<br>  }<br>  res.end(JSON.stringify({<br>    url: req.url,<br>    pid: process.pid,<br>  }));<br>});<br><br>module.exports = server;</code></pre><br><br><h3><a href="https://github.com/aleafs/pm/blob/master/demo/graceful_exit/worker.js">worker.js</a></h3><br><br><pre><code>var graceful = require(&#39;graceful&#39;);<br>var worker = require(&#39;pm&#39;).createWorker();<br>var server = require(&#39;./app&#39;);<br><br>// hack for pm, because server.<em>handle is empty.<br>server.close = function () {};<br><br>graceful({<br>  server: server,<br>  worker: worker,<br>  error: function (err) {<br>    console.log(&#39;[%s] [worker:%s] error: %s&#39;, new Date(), process.pid, err.stack);<br>  },<br>  killTimeout: 10000,<br>});<br><br>worker.ready(function (socket, port) {<br>  server.emit(&#39;connection&#39;, socket);<br>});</code></pre><br><br><h3>stdout</h3><br><br><pre><code>$ node demo/graceful_exit/dispatch.js <br>[Fri Apr 19 2013 09:28:38 GMT+0800 (CST)] [master:9773] new web:worker:9774 fork<br>[Fri Apr 19 2013 09:28:38 GMT+0800 (CST)] [master:9773] new web:worker:9775 fork<br><br>$ curl localhost:1984/<br>{&quot;url&quot;:&quot;/&quot;,&quot;pid&quot;:9775}<br><br>$ curl localhost:1984/asyncerror<br># hang for a while<br>curl: (52) Empty reply from server<br><br># Master<br>$ node demo/graceful_exit/dispatch.js <br>[Fri Apr 19 2013 09:28:38 GMT+0800 (CST)] [master:9773] new web:worker:9774 fork<br>[Fri Apr 19 2013 09:28:38 GMT+0800 (CST)] [master:9773] new web:worker:9775 fork<br>[Fri Apr 19 2013 09:29:35 GMT+0800 (CST)] [worker:9775] error: ReferenceError: asyncError is not defined<br>    at Object.</em>onTimeout (/Users/mk2/git/pm/demo/graceful_exit/app.js:18:7)<br>    at Timer.list.ontimeout (timers.js:101:19)<br>[uncaughtException] throw error 1 times<br>[ReferenceError: asyncError is not defined]<br>[Fri Apr 19 2013 09:29:35 GMT+0800 (CST)] [worker:9775] close 1 servers!<br>[Fri Apr 19 2013 09:29:35 GMT+0800 (CST)] [worker:9775] worker disconnect!<br>[Fri Apr 19 2013 09:29:35 GMT+0800 (CST)] [master:9773] new web:worker:10089 fork<br>[Fri Apr 19 2013 09:29:35 GMT+0800 (CST)] [master:9773] worker:9775 disconnect! new worker:10089 fork<br>[Fri Apr 19 2013 09:29:45 GMT+0800 (CST)] [worker:9775] kill timeout, exit now.<br>[Fri Apr 19 2013 09:29:45 GMT+0800 (CST)] [master:9773] web:worker:9775 quit, code: 1, signal: null<br><br># test again<br>$ curl localhost:1984/asyncerror<br># hang for a while<br>curl: (52) Empty reply from server<br><br># Master<br>$ node demo/graceful_exit/dispatch.js <br>[Fri Apr 19 2013 09:28:38 GMT+0800 (CST)] [master:9773] new web:worker:9774 fork<br>[Fri Apr 19 2013 09:28:38 GMT+0800 (CST)] [master:9773] new web:worker:9775 fork<br>[Fri Apr 19 2013 09:29:35 GMT+0800 (CST)] [worker:9775] error: ReferenceError: asyncError is not defined<br>    at Object.<em>onTimeout (/Users/mk2/git/pm/demo/graceful_exit/app.js:18:7)<br>    at Timer.list.ontimeout (timers.js:101:19)<br>[uncaughtException] throw error 1 times<br>[ReferenceError: asyncError is not defined]<br>[Fri Apr 19 2013 09:29:35 GMT+0800 (CST)] [worker:9775] close 1 servers!<br>[Fri Apr 19 2013 09:29:35 GMT+0800 (CST)] [worker:9775] worker disconnect!<br>[Fri Apr 19 2013 09:29:35 GMT+0800 (CST)] [master:9773] new web:worker:10089 fork<br>[Fri Apr 19 2013 09:29:35 GMT+0800 (CST)] [master:9773] worker:9775 disconnect! new worker:10089 fork<br>[Fri Apr 19 2013 09:29:45 GMT+0800 (CST)] [worker:9775] kill timeout, exit now.<br>[Fri Apr 19 2013 09:29:45 GMT+0800 (CST)] [master:9773] web:worker:9775 quit, code: 1, signal: null<br><br>[Fri Apr 19 2013 09:30:29 GMT+0800 (CST)] [worker:9774] error: ReferenceError: asyncError is not defined<br>    at Object.</em>onTimeout (/Users/mk2/git/pm/demo/graceful_exit/app.js:18:7)<br>    at Timer.list.ontimeout (timers.js:101:19)<br>[uncaughtException] throw error 1 times<br>[ReferenceError: asyncError is not defined]<br>[Fri Apr 19 2013 09:30:29 GMT+0800 (CST)] [worker:9774] close 1 servers!<br>[Fri Apr 19 2013 09:30:29 GMT+0800 (CST)] [worker:9774] worker disconnect!<br>[Fri Apr 19 2013 09:30:29 GMT+0800 (CST)] [master:9773] new web:worker:10119 fork<br>[Fri Apr 19 2013 09:30:29 GMT+0800 (CST)] [master:9773] worker:9774 disconnect! new worker:10119 fork<br>[Fri Apr 19 2013 09:30:39 GMT+0800 (CST)] [worker:9774] kill timeout, exit now.<br>[Fri Apr 19 2013 09:30:39 GMT+0800 (CST)] [master:9773] web:worker:9774 quit, code: 1, signal: null</code></pre><br><br><h2>展望</h2><br><br><p>之前想依赖 [domain] 实现 <code>uncaughtException</code> 发生的时候，给当前请求响应友好的500错误，现在看来是无法100%实现的。</p><br><br><p>而且引入 [domain] 会造成一定的性能损耗 <a href="http://fengmk2.github.io/blog/2013/03/domain-helloworld-benchmark.html">nodejs domain module hello world and benchmark</a>:</p><br><br><ul><br><li>normal: <code>7624.59 trans/sec</code></li><br><li>domain: <code>7068.83 trans/sec</code></li><br></ul><br><br><p>所以，按目前的情况还是不使用 [domain] ，然后优雅退出解决未捕获异常发生后可能出现的一切问题。</p><img src="http://www.cnblogs.com/fengmk2/aggbug/3123511.html?type=1" width="1" height="1" alt=""><p><a href="http://www.cnblogs.com/fengmk2/archive/2013/06/07/graceful-exit-pm-cluster.html">本文链接</a></p></p>

</div>


            <div class="footer">
              <div class="contact">
                <p>
                  Your Name<br />
                  What You Are<br />
                  your@email.com
                </p>
              </div>
              <div class="contact">
                <p>
                  <a href="http://github.com/lyuehh/">github.com/lyuehh</a><br />
                  <a href="http://twitter.com/lyuehh/">twitter.com/lyuehh</a><br />
                </p>
              </div>
            </div>
          </div>
        </div> <!-- /container -->

    </body>
</html>
